std/array.d

```diff
@@ -3519,6 +3519,74 @@ if (isDynamicArray!A)
         return cast(typeof(return))(_data ? _data.arr : null);
     }
 
+    /**
+     * Returns: The current length of the array
+     */
+    @property size_t length() const
+    {
+        return _data ? _data.arr.length : 0U;
+    }
+
+    /**
+     * Set a new length for the array
+     * Params:
+     *    a = The new length
+     */
+    @property void length(size_t newLength) @trusted
```

Comment:
What's the use case?

Reply:
Behaving like a dynamic array while expanding/shrinking it like an appender. There is no proper way to expand with the appender factor while setting initialized memory.

Reply:
But it's not a dynamic array - it's a tool for building one. I don't understand what the use case would be.

Reply:
For building arrays with the help of other language APIs (i.e. C). Reserve a capacity, then feed the pointer/length pair to an API to fill it, then set the length. Can't be `@trusted` though like it seems to be now.

Reply:
> But it's not a dynamic array - it's a tool for building one. I don't understand what the use case would be.

It should mimic a dynamic array. The appender is literally used as an optimization to `~=` operator that mimics a dynamic array but with a larger factor for `.capacity` to do fewer allocations. There are also other things missing in the Appender like other ranges interfaces, but not related to this PR. I see no reasons to restrict this interface to an OutputRange. It makes no sense to me that people need to use `opSlice` to do everything else, making it a special case.

Reply:
> For building arrays with the help of other language APIs (i.e. C). Reserve a capacity, then feed the pointer/length pair to an API to fill it, then set the length. Can't be `@trusted` though like it seems to be now.

Can you elaborate? I didn't understand your point.

Reply:
Why should it mimic a dynamic array?

Reply:
> Can you elaborate? I didn't understand your point.

'Twas a reply to Atila's question about the use case :)

Reply:
> Why it shouldn't?

The default response to adding features should be to say no unless there's a compelling argument that said feature has a positive ROI. 

> Just making it hard and another special case

Could you please explain what's hard about creating the dynamic array then using its length?

> Users literally use Appender just to reduce consecutive allocations

Yes. Hopefully after having used a profiler, otherwise they should use dynamic arrays instead.

> except for now that people need to do [] every time to access the enclosing dynamic array

This is not how Appender is meant to be used which is to build an array then return it.

Reply:
> Because we already have dynamic arrays built into the language. Introducing another implementation of the exact same thing in Phobos adds to our maintenance burden for no benefit.
>
> The fact that `doInitialize` is literally [copy-pasted from druntime](https://github.com/dlang/druntime/blob/cd0806a7f94d1fca85bf9fcc436c9d43424f26e3/src/rt/lifetime.d#L1734-L1749) should be a red flag indicating that you have made a questionable decision somewhere along the line here.

I'm not talking about reimplementing it, I'm just suggesting having the same interface. The maintenance burden is zero if the duplicated stuff gets refactored afterwards, hence "copy-pasted". The compiler hook for increasing array capacity needs a refactor anyway since it uses `TypeInfo`, and it shouldn't because it is now templatized. After a refactoring to a `unsafeEmplace` internals, this shouldn't be an issue. To clarify, this is not the only place where it is needed. See `Array` container. It uses `initializeAll` which is not only inefficient in this case (reserve does already zero init) but it also doesn't cover the case of `arr.length = ` when structs add `@disable this(this);`. I can add that to my TODO list tho, but this is to prove that there is a need for some refactoring here.

> Could you please explain what's hard about creating the dynamic array then using its length?

Currently, Appender only has `reserve` and that's the only way to increase its capacity with the custom capacity factor of the Appender. Again, the reason to have an appender in the first place is to have a different capacity factor, when reallocating, so it can do less allocations. In normal dynamic arrays `.length` makes it possible to do similar to `.reserve()` but emplace the reserved space (minus extra space managed by capacity) with the `.init` value efficiently.

> Yes. Hopefully after having used a profiler, otherwise they should use dynamic arrays instead.

Sometimes there is no need to use a profiler for extremely obvious cases:
```d
string[] a;
foreach (str; strange.byLine())
{
	// ...
	a ~= str;
	// ...
}
```

If I know that `strange` can have, let's say, around `1000000` lines, looking at the code you can clearly see the issue. Now let's say that I use `a` somewhere else in the code that uses the normal dynamic array interface. Great, you will have to change every occurrence of `a`.

---

> This is not how Appender is meant to be used which is to build an array then return it.

Your argument here is: Appender is not meant to be that way but I fail to understand why. What is the reason to make it a different interface? Performance? No, performance is exactly the same. Safeness? No, everything is allocated with the GC, so memory ownership is not an issue, and even if it was, those methods would be marked as `@system`.

There are other interfaces that mimic/implement a range interface that doesn't sound like a range or not "meant to be" but might be very handy for the user. This includes, e.g. this change https://github.com/dlang/phobos/pull/8417 . In this case, it doesn't even apply because Appender is literally a wrapper of a dynamic array.

> > Why it shouldn't?
>
> The default response to adding features should be to say no unless there's a compelling argument that said feature has a positive ROI.

I tried to answer why it should but you didn't make an effort to have a rationale why it shouldn't other than a no.

Reply:
> Again, the reason to have an appender in the first place is to have a different capacity factor, when reallocating, so it can do less allocations.

According to [the documentation][1], the reason to have an appender is to make repeated use of `~=` more efficient by storing the capacity locally (rather than in the global heap metadata, where accessing it requires synchronization). Both Appender and built-in dynamic arrays grow exponentially, so the number of allocations should not differ substantially between the two.

If you are going to reserve all of the memory you need up front and then fill it in, there is no benefit to using an appender over a built-in dynamic array. (In fact, there is a small penalty, since Appender allocates its capacity metadata on the heap.)

[1]: https://phobos.dpldocs.info/std.array.Appender.html

Reply:
> > Again, the reason to have an appender in the first place is to have a different capacity factor, when reallocating, so it can do less allocations.
> 
> According to [the documentation](https://phobos.dpldocs.info/std.array.Appender.html), the reason to have an appender is to make repeated use of `~=` more efficient by storing the capacity locally (rather than in the global heap metadata, where accessing it requires synchronization). Both Appender and built-in dynamic arrays grow exponentially, so the number of allocations should not differ substantially between the two.

That's not the only differentiating factor of an Appender.

It does differ substantially between the two, not only in the number of reallocations but also in the number of bytes reallocated. There is a different growing scheme for efficient appending other than `rt.lifetime.newCapacity`. See https://github.com/dlang/phobos/blob/ea92d8e8067e08934dfaab92175624ae694d9fea/std/array.d#L3961 . You can also run and see by yourself with `tracegc`. `rt.lifetime.newCapacity` only tries to exponentially set capacity beyond memory page size. I also think that we should study about increasing the factor further with some performance tests, specifically for the appender.

> If you are going to reserve all of the memory you need up front and then fill it in, there is no benefit to using an appender over a built-in dynamic array. (In fact, there is a small penalty, since Appender allocates its capacity metadata on the heap.)

I mentioned a case where it can be a huge amount of allocations, but not necessarily. The use cases where you use an allocator are obviously when you don't have the possibility of preallocating, or if you do, worst-case scenario, you will waste time allocating a huge block of memory for a very small portion of data.

Reply:
I think we are deviating from the point tho. Let's say the major factor is synchronization or growth rate, doesn't really matter. The user may want to shift from a dynamic array to an appender, for whatever reason. What is the reason for the user to require amending all the occurrences of the array and use `opSlice`, instead of having a convenient way?

Reply:
> The user may want to shift from a dynamic array to an appender

Then they're not using appender the way it was intended.

> What is the reason for the user to require amending all the occurrences of the array and use opSlice,

That wouldn't be the fix. Appender is meant to be used to construct an array, so the fix is to do that and return that array where it'll be used as normal.

Reply:
> > What is the reason for the user to require amending all the occurrences of the array and use opSlice,
> 
> That wouldn't be the fix. Appender is meant to be used to construct an array, so the fix is to do that and return that array where it'll be used as normal.

The fact that the appender is meant to be used to construct/append to an array doesn't mean that it shouldn't have the other normal interfaces. It is literally interfaces. I don't see drawbacks to implementing common/stable/standard array interfaces that doesn't hurt on performance.

This interface is, in fact, the most concerned one because it behaves differently due to capacity factor and there is no easy and safe way to implement it.

Tell me a way the user can do the following:

```d
Appender!(int[]) a;
// other append operations

// I want to increase the rest of the space with initialized memory
foreach(a; iota(restOfTheSpace))
	a ~= int.init;
```

This need to be:
- efficiently without accessing an external capacity
- be able to set struct with `this(this) @disable`
- compatible with dynamic array interfaces

The only way I see this fitting in both cases is:

```d
Appender!(int[]) a;
// other append operations

// I want to increase the rest of the space with initialized memory
auto len = a[].length;
a.reserve(restOfTheSpace);
initializeAll(a[][len .. $]); // if `this(this) @disable` get fixed
```

This case can be problematic if a Range is given with opSlice interface, but the opSlice is expensive.

The way I'm proposing is a more natural way, either it's a dynamic array:
```d
int[] a;
// other append operations

a.length = restOfTheSpace;
```

or an `Appender`:
```d
Appender!(int[]) a;
// other append operations

a.length = restOfTheSpace;
```

To complete my point on the growth rate, run the following:

```d
void main()
{
    Appender!(string[]) ap;
    string[] ar;
    foreach(_; 0 .. 30)
    {
        writeln("ap: ", ap.capacity);
        writeln("ar: ", ar.capacity);
        ap.reserve(ap.capacity + 1);
        ar.reserve(ar.capacity + 1);
    }
}
```

or
```d
void main()
{
    Appender!(ubyte[]) ap;
    ubyte[] ar;
    foreach(_; 0 .. 30)
    {
        writeln("ap: ", ap.capacity);
        writeln("ar: ", ar.capacity);
        ap.reserve(ap.capacity + 1);
        ar.reserve(ar.capacity + 1);
    }
}
```

to see the different growing rates of a dynamic array and an appender.

---

It is also hacky to support other standard array/range ops, making it pretty much a special case. I see that your argument is biased to "it is not meant to be that way" without further ellaboration and it is unfortunate that I can't find any other reason. I would like to see an ellaborated version of your argument and a list of drawbacks that such addition would have. Although, if you don't see practical value on this anyway, I might just close this and move on.

Reply:
> The fact that the appender is meant to be used to construct/append to an array doesn't mean that it shouldn't have the other normal interfaces

I disagree.

> I don't see drawbacks to implementing common/stable/standard array interfaces

As mentioned earlier, the drawback is the existence of the code given that, in my opinion, it doesn't justify itself. Worse, it incentivises using the appenders as arrays.

Reply:
> As mentioned earlier, the drawback is the existence of the code given that, in my opinion, it doesn't justify itself. Worse, it incentivizes using the appenders as arrays.

I agree with this. I don't think `Appender` should be used as an array at all. I think the problem here is the design approach of it, as seen by your interpretation. An appender should be no more no less than a mechanism to append, not an array. The issue is that by this design the user is inclined to think otherwise since he creates a new instance from an array, which gives the illusion that it might be usable as an array itself. IMO the correct way to view the `Appender` is to look at it as some sort of policy that alters the way memory usage is made while appending. So maybe it's the design of it that's misleading.

Maybe a more explicit approach to what `Appender` is would be to make appending functions take an appender or some sort of policy plus a predicate implementing the appending logic. That way it would be more explicit the meaning of it and would ensure the policy is temporary, as it should be used that way.

Reply:
> I agree with this. I don't think `Appender` should be used as an array at all. I think the problem here is the design approach of it, as seen by your interpretation. An appender should be no more no less than a mechanism to append, not an array. The issue is that by this design the user is inclined to think otherwise since he creates a new instance from an array, which gives the illusion that it might be usable as an array itself. IMO the correct way to view the `Appender` is to look at it as some sort of policy that alters the way memory usage is made while appending. So maybe it's the design of it that's misleading.

Designing an API to be explicitly restrictive because it is the only intended purpose is not a generic approach to be on a modern standard library. Taking the same example of `Nullable`, we could just strictly implement the methods needed to make `nullable` work as an actual `nullable` and not add those range interfaces, but is it handy for the user? It adds value and doesn't remove value.

I can't understand the described practical drawbacks of such addition:

The idea of misusing `Appender` as an array seems incorrect. No one should use it like a normal array unless it fulfils the documented use-cases, which are the characteristics that make it different from a normal array. You can argue that I can always use a `DList` versus a `SList` because `DList` does the same things as `SList`, but that doesn't mean it is always adapted for your use-case. You can also do other comparisons with other data structures. For me, arguing that is not a valid reason to explicitly restrict an interface.

> Maybe a more explicit approach to what `Appender` is would be to make appending functions take an appender or some sort of policy plus a predicate implementing the appending logic. That way it would be more explicit the meaning of it and would ensure the policy is temporary, as it should be used that way.

Having `append` functions that uses `Appender` struct would make a local `capacity` field useless, hence it needs to keep its context, otherwise, the performance would be always worse. This is due to how `Appender` and dynamic arrays capacity are implemented, otherwise, a bigger growing factor would suffice.

---

My vision of these changes is ultimately have a configurable `Array` container with a `Grower` hook or something similar and maybe other synchronisation features, apart from `Appender` interface, but the containers on Phobos doesn't seem to follow the range interface anyway, hence community often use emsi containers along with other features.

I see no productiveness/progress on continuing arguing here, so I'm going to close this for now.